FEATURE_SECURE_PROCESSING threshold too low in XmlDecoder

Got:
> <=<=Exception in thread "main" org.metafacture.framework.MetafactureException: org.xml.sax.SAXParseException; lineNumber: 8496675; columnNumber: 3876; JAXP00010004: Die akkumulierte Größe von Entitys ist "50.000.001" und überschreitet den Grenzwert "50.000.000", der von "FEATURE_SECURE_PROCESSING" festgelegt wurde.
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:79)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:44)
        at org.metafacture.io.FileOpener.process(FileOpener.java:158)
        at org.metafacture.io.FileOpener.process(FileOpener.java:41)
        at org.metafacture.flux.parser.StringSender.process(StringSender.java:43)
        at org.metafacture.flux.parser.Flow.start(Flow.java:118)
        at org.metafacture.flux.parser.FluxProgramm.start(FluxProgramm.java:168)
        at org.metafacture.runner.Flux.main(Flux.java:87)
Caused by: org.xml.sax.SAXParseException; lineNumber: 8496675; columnNumber: 3876; JAXP00010004: Die akkumulierte Größe von Entitys ist "50.000.001" und überschreitet den Grenzwert "50.000.000", der von "FEATURE_SECURE_PROCESSING" festgelegt wurde.
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1243)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:635)
        at org.metafacture.xml.XmlDecoder.process(XmlDecoder.java:73)

when analyzing Alma Basedump in a single process.

--- a/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
+++ b/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
@@ -48,18 +48,34 @@ import javax.xml.parsers.SAXParserFactory;
 public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
 
     private static final String SAX_PROPERTY_LEXICAL_HANDLER = "http://xml.org/sax/properties/lexical-handler";
-
-    private final XMLReader saxReader;
+    private XMLReader saxReader;
+    private final SAXParserFactory parserFactory = SAXParserFactory.newInstance();
 
     /**
-     * Constructs an XmlDecoder by obtaining a new instance of an
+     * Creates an instance of {@link XmlDecoder} by obtaining a new instance of an
      * {@link org.xml.sax.XMLReader}.
      */
     public XmlDecoder() {
         try {
-            final SAXParserFactory parserFactory = SAXParserFactory.newInstance();
             parserFactory.setNamespaceAware(true);
+            saxReader = parserFactory.newSAXParser().getXMLReader();
+        }
+        catch (final ParserConfigurationException | SAXException e) {
+            throw new MetafactureException(e);
+        }
+    }
 
+    /**
+     * Sets the total entity size limit for the XML parser.
+     * See <a href="https://docs.oracle.com/en/java/javase/13/security/java-api-xml-processing-jaxp-security-guide.html#GUID-82F8C206-F2DF-4204-9544-F96155B1D258__TABLE_RQ1_3PY_HHB">java-api-xml-processing-jaxp-security-guide.html</a>
+     *
+     * Defaults to "50,000,000". Set to "0" to allow unlimited entities.
+     *
+     * @param size the size of the allowed entities. Set to "0" if entities should be unlimited.
+     */
+    public void setTotalEntitySizeLimit(final String size) {
+        try {
+            System.setProperty("jdk.xml.totalEntitySizeLimit", size);
             saxReader = parserFactory.newSAXParser().getXMLReader();
         }
         catch (final ParserConfigurationException | SAXException e) {

--- a/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
+++ b/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
@@ -88,10 +88,7 @@ public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
         try {
             saxReader.parse(new InputSource(reader));
         }
-        catch (final IOException e) {
-            throw new MetafactureException(e);
-        }
-        catch (final SAXException e) {
+        catch (final IOException | SAXException e) {
             throw new MetafactureException(e);
         }
     }
@@ -105,10 +102,7 @@ public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
         try {
             saxReader.setProperty(SAX_PROPERTY_LEXICAL_HANDLER, getReceiver());
         }
-        catch (final SAXNotRecognizedException e) {
-            throw new MetafactureException(e);
-        }
-        catch (final SAXNotSupportedException e) {
+        catch (final SAXNotRecognizedException | SAXNotSupportedException e) {
             throw new MetafactureException(e);
         }
     }

--- a/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
+++ b/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
@@ -41,7 +41,7 @@ import javax.xml.parsers.SAXParserFactory;
  * @author Christoph Böhme
  *
  */
-@Description("Reads an XML file and passes the XML events to a receiver.")
+@Description("Reads an XML file and passes the XML events to a receiver. Set 'totalEntitySizeLimit=\"0\"' to allow unlimited XML entities.")
 @In(Reader.class)
 @Out(XmlReceiver.class)
 @FluxCommand("decode-xml")

--- a/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
+++ b/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
@@ -48,8 +48,8 @@ import javax.xml.parsers.SAXParserFactory;
 public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
 
     private static final String SAX_PROPERTY_LEXICAL_HANDLER = "http://xml.org/sax/properties/lexical-handler";
-    private XMLReader saxReader;
-    private final SAXParserFactory parserFactory = SAXParserFactory.newInstance();
+    private static final String TOTAL_ENTITY_SIZE_LIMIT = "http://www.oracle.com/xml/jaxp/properties/totalEntitySizeLimit";
+    private final XMLReader saxReader;
 
     /**
      * Creates an instance of {@link XmlDecoder} by obtaining a new instance of an
@@ -57,6 +57,7 @@ public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
      */
     public XmlDecoder() {
         try {
+            final SAXParserFactory parserFactory = SAXParserFactory.newInstance();
             parserFactory.setNamespaceAware(true);
             saxReader = parserFactory.newSAXParser().getXMLReader();
         }
@@ -71,14 +72,13 @@ public final class XmlDecoder extends DefaultObjectPipe<Reader, XmlReceiver> {
      *
      * Defaults to "50,000,000". Set to "0" to allow unlimited entities.
      *
-     * @param size the size of the allowed entities. Set to "0" if entities should be unlimited.
+     * @param totalEntitySizeLimit the size of the allowed entities. Set to "0" if entities should be unlimited.
      */
-    public void setTotalEntitySizeLimit(final String size) {
+    public void setTotalEntitySizeLimit(final String totalEntitySizeLimit) {
         try {
-            System.setProperty("jdk.xml.totalEntitySizeLimit", size);
-            saxReader = parserFactory.newSAXParser().getXMLReader();
+            saxReader.setProperty(TOTAL_ENTITY_SIZE_LIMIT, totalEntitySizeLimit);
         }
-        catch (final ParserConfigurationException | SAXException e) {
+        catch (final SAXException e) {
             throw new MetafactureException(e);
         }
     }

--- a/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
+++ b/metafacture-xml/src/main/java/org/metafacture/xml/XmlDecoder.java
@@ -41,7 +41,7 @@ import javax.xml.parsers.SAXParserFactory;
  * @author Christoph Böhme
  *
  */
-@Description("Reads an XML file and passes the XML events to a receiver. Set 'totalEntitySizeLimit=\"0\"' to allow unlimited XML entities.")
+@Description("Reads an XML file and passes the XML events to a receiver. Set `totalEntitySizeLimit=\"0\"` to allow unlimited XML entities.")
 @In(Reader.class)
 @Out(XmlReceiver.class)
 @FluxCommand("decode-xml")

--- a/metafacture-xml/src/test/java/org/metafacture/xml/XmlDecoderTest.java
+++ b/metafacture-xml/src/test/java/org/metafacture/xml/XmlDecoderTest.java
@@ -1,0 +1,69 @@
+/*
+ * Copyright 2024 Pascal Christoph (hbz)
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.metafacture.xml;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.metafacture.framework.MetafactureException;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+
+/**
+ * Tests for class {@link XmlDecoder}.
+ *
+ * @author Pascal Christoph (dr0i)
+ */
+public final class XmlDecoderTest {
+
+    private final String TEST_XML_WITH_TWO_ENTITIES = "<record>&gt;&gt;</record>";
+    private XmlDecoder xmlDecoder;
+    private final Reader reader = new StringReader(TEST_XML_WITH_TWO_ENTITIES);
+
+    @Before
+    public void initSystemUnderTest() {
+        xmlDecoder = new XmlDecoder();
+    }
+
+    @Test
+    public void issue554_default() {
+        process(xmlDecoder);
+    }
+
+    @Test(expected = MetafactureException.class)
+    public void issue554_shouldFail() {
+        xmlDecoder.setTotalEntitySizeLimit("1");
+        process(xmlDecoder);
+    }
+
+    @Test
+    public void issue554_unlimitedEntities() {
+        xmlDecoder.setTotalEntitySizeLimit("0");
+        process(xmlDecoder);
+    }
+
+    private void process(XmlDecoder xmlDecoder) {
+        try {
+            xmlDecoder.process(reader);
+            reader.close();
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

